luarules/gadgets/api_unit_suspension.lua

```diff
@@ -0,0 +1,556 @@
+	GG.SuspendMovement               = nil
+	GG.ResumeMovement                = nil
+	GG.GetUnitIsSuspendedMoveCtrl    = nil
+end
+
+if suspendMovement then
+	function gadget:GameFrame(frame)
+		-- For units sent flying that halt, trapped in void that recover, etc:
+		for unitID in pairs(postponedMoveCtrl) do
+			if not canMoveCtrl(unitID) then
+				postponedMoveCtrl[unitID] = nil
+				stopMoving(unitID)
+			end
+		end
+	end
+end
+
+function gadget:UnitDestroyed(unitID, unitDefID, unitTeam, attackerID, attackerDefID, attackerTeam, weaponDefID)
+	suspendedUnits[unitID] = nil
+	suspendedMoveCtrl[unitID] = nil
+	postponedMoveCtrl[unitID] = nil
+end
+
+-- Removing suspensions --------------------------------------------------------
+
+for suspend, resume in pairs(suspendReasons) do
+	gadget[resume] = function(_, unitID)
+		if suspendedUnits[unitID] ~= nil then
+			clearSuspendReason(unitID, resume)
+		end
+	end
+end
+
+-- One of these generic callins, UnitStunned, works differently:
+
+function gadget:UnitStunned(unitID, unitDefID, unitTeam, stunned)
+	if stunned then
+		addSuspendReason(unitID, "UnitStunned", false)
+	elseif suspendedUnits[unitID] then
+		clearSuspendReason(unitID, "UnitStunned")
+	end
+end
+
+-- Suspension behaviors --------------------------------------------------------
+
+function gadget:AllowCommand(unitID, unitDefID, unitTeam, cmdID, cmdParams, cmdOptions, cmdTag)
+	return not suspendedUnits[unitID] or not commandSuspendDisallows[cmdID]
+end
+
+-- Use the Allow* callins to fail attempts at tasks as early as possible,
+-- so that other gadgets can handle the failure cases as early as possible.
+
+local function shouldAllow(_, unitID)
+	return suspendedUnits[unitID] == nil
+end
+
+if commandSuspendDisallows[CMD.ATTACK] then
+	gadget.AllowWeaponTargetCheck = shouldAllow
+	gadget.AllowUnitKamikaze = shouldAllow
+end
+
+if commandSuspendDisallows[CMD.CLOAK] then
+	-- Note: We don't actually remove the cloaked state (or clear any state).
+	gadget.AllowUnitCloak = shouldAllow
+end
+
+if commandSuspendDisallows[CMD.LOAD_UNITS] and commandSuspendDisallows[CMD.UNLOAD_UNIT] then
+	gadget.AllowUnitTransport = shouldAllow
+elseif commandSuspendDisallows[CMD.LOAD_UNITS] then
+	gadget.AllowUnitTransportLoad = shouldAllow
+elseif commandSuspendDisallows[CMD.UNLOAD_UNIT] then
+	gadget.AllowUnitTransportUnload = shouldAllow
+end
+
+if commandSuspendDisallows[CMD.CAPTURE] then
+	gadget.AllowUnitCaptureStep = shouldAllow
+end
+
+if commandSuspendDisallows[CMD.RECLAIM] and commandSuspendDisallows[CMD.REPAIR] then
+	gadget.AllowUnitBuildStep = shouldAllow
+elseif commandSuspendDisallows[CMD.RECLAIM] then
+	function gadget:AllowUnitBuildStep(builderID, builderTeam, featureID, featureDefID, part)
+		return part > 0 or suspendedUnits[unitID] == nil
+	end
+elseif commandSuspendDisallows[CMD.RESURRECT] then
+	function gadget:AllowUnitBuildStep(builderID, builderTeam, featureID, featureDefID, part)
+		return part < 0 or suspendedUnits[unitID] == nil
+	end
```

Comment:
these say UnitBuildStep but there's featureID. inside it also references unitID.